DeepCSV and CSVv2 negative tagger bugfix

emilbols opened the pull request:
A series of bugs that were discovered in the current implementation of the DeepCSV and CSVv2 negative tagger has been fixed in this PR. A description of the bugs and the negative taggers in general can be found in this presentation at the BTV meeting [1].

It was also discovered that in the method used for producing inclusive secondary vertices, there is a cut on the angle between the vertex flight direction and the direction of the track sum. This cut needs to be inverted for a negative tagger, so a vertexing sequence with correct cuts for a negative tagger has been introduced.

In the plot [2] the negative tagger of DeepCSV is shown. The negative tagger discriminator values are shown from -1 to 0, and the full tagger is shown from 0 to 1.

[1] https://indico.cern.ch/event/713923/contributions/2938152/attachments/1620844/2578707/NegativeTagger.pdf

[2]
![deepcsvnegativetagger](https://user-images.githubusercontent.com/33452471/40061942-5945600a-585a-11e8-9b78-1286ab57791a.png)

emilbols committed:
Fix of the DeepCSV Negative Tagger, and introduction of DeepFlavour Negative Tag definition

emilbols committed:
Use correct vertex collection for DeepFlavour negative tagger

emilbols committed:
Changing vertex cuts for negative tagger vertex reconstruction

emilbols committed:
Added correct vertex collection to DeepCSV+DeepFlavour

emilbols committed:
Flipping mass above charm correctly

emilbols committed:
formatting

emilbols committed:
format

emilbols committed:
Merge pull request #1 from emilbols/from-CMSSW_10_1_0-NegativeTaggerUpdates

From cmssw 10 1 0 negative tagger updates

emilbols committed:
Updated the DeepFlavour model to the newest training.

emilbols committed:
Negative tagger updates

cmsbuild commented:
The code-checks are being triggered in jenkins.

emilbols committed:
code style

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23206/4665

cmsbuild commented:
A new Pull Request was created by  @emilbols  for master.

It involves the following packages:

PhysicsTools/PatAlgos
RecoBTag/Combined
RecoBTag/DeepFlavour
RecoBTag/SecondaryVertex
RecoVertex/AdaptiveVertexFinder


@perrotta, @monttj, @cmsbuild, @slava77, @gpetruc, @arizzi can you please review it and eventually sign? Thanks.
@TaiSakuma, @gouskos, @felicepantaleo, @rappoccio, @ahinzmann, @seemasharmafnal, @mmarionncern, @imarches, @makortel, @acaudron, @jdolen, @ferencek, @GiacomoSguazzoni, @rovere, @VinInn, @jdamgov, @nhanvtran, @gkasieczka, @schoef, @ebrondol, @dgulhan, @JyothsnaKomaragiri, @mverzett, @gpetruc, @mariadalfonso, @pvmulder this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@emilbols @jmduarte 
this and #23033 overlap significantly and can not be merged or reviewed independently.
Please clarify if there is some priority preference.
By default, it's a FIFO.

slava77 commented:
@cmsbuild please test with cms-sw/cmsdist#3944

cmsbuild commented:
The tests are being triggered in jenkins.
Using externals from cms-sw/cmsdist#3944

cmsbuild commented:
+1
Tested at: 36b1cc67e21144c9139671d4155093aa403c49b7
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23206/27972/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/f519fc5608cb96c92b5768cb699fd1fcd5314cdb
https://github.com/cms-sw/cmssw/commit/7f15f5baf91fbddecf178649162f80d435067acd
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23206/27972/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23206/27972/git-merge-result

cmsbuild commented:
Comparison job queued.

imarches commented:
The understanding is that the DeepDoubleB team has still some checks ongoing (@jmduarte  may confirm), so probably this PR could go in first, given also that it does include major developments and it may be potentially be reviewed rather fast.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23206/27972/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 108 differences found in the comparisons
 - DQMHistoTests: Total files compared: 30
 - DQMHistoTests: Total histograms compared: 2740106
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2739922
 - DQMHistoTests: Total skipped: 183
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 29 files compared)
 - Checked 124 log files, 14 edm output root files, 30 DQM output files

jmduarte commented:
@cvernier @ferencek @imarches @slava77 Yes, we are still figuring out a (small) configuration issue in our PR #23033, but this PR introduces some important changes to `SVConverter.h`, `ChargedCandidateConverter.h`, `DeepFlavourTFJetTagsProducer.cc`, and `DeepFlavourTagInfoProducer.cc` which will be very difficult for us to rebase to and merge with (given our changes and complete renaming of the same files as requested by @slava77)

Maybe we can discuss with the authors of this PR the best way to proceed?

Thanks,

Javier

slava77 commented:
On 5/18/18 12:47 AM, Javier Duarte wrote:
> @cvernier <https://github.com/cvernier> @ferencek
> <https://github.com/ferencek> @imarches <https://github.com/imarches>
> @slava77 <https://github.com/slava77> Yes, we are still figuring out a
> (small) configuration issue in our PR #23033
> <https://github.com/cms-sw/cmssw/pull/23033>, but this PR introduces
> some important changes to |SVConverter.h|,
> |ChargedCandidateConverter.h|, |DeepFlavourTFJetTagsProducer.cc|, and
> |DeepFlavourTagInfoProducer.cc| which will be very difficult for us to
> rebase to and merge with (given our changes and complete renaming of the
> same files as requested by @slava77 <https://github.com/slava77>)
> 
> Maybe we can discuss with the authors of this PR the best way to proceed?

sure.
this was the reason for my earlier question about PR ordering

> 
> Thanks,
> 
> Javier
>

imarches commented:
Hi @jmduarte for me each ordering is fine, provided that the DeepDoubleB PR will not take too long making this PR miss the 10_2 deadlines. It is mandatory to have this merged in time for 10_2.

emilbols commented:
@imarches @slava77 @jmduarte 

Its fine for me, if https://github.com/cms-sw/cmssw/pull/23033 is going first. It will not be a lot of work for me to reapply my changes on top of that PR (i hope).

slava77 commented on RecoBTag/DeepFlavour/plugins/ChargedCandidateConverter.h:
```diff
@@ -30,7 +30,19 @@ namespace btagbtvdeep {
+                                            ChargedCandidateFeatures & c_pf_features,
+					    const bool flip) {
+
+	float TrackSip2dVal = track_info.getTrackSip2dVal();
```
start local variables and data members with lower case letter. Capitalization is reserved for class names.

emilbols replied:
@slava77 Done in the new PR at https://github.com/cms-sw/cmssw/pull/23467/files#diff-898f4df86b3e380a58ff390ca3e02194R23

slava77 commented on RecoBTag/DeepFlavour/plugins/ChargedCandidateConverter.h:
```diff
@@ -30,7 +30,19 @@ namespace btagbtvdeep {
+	float TrackSip3dVal = track_info.getTrackSip3dVal();
+	float TrackSip3dSig = track_info.getTrackSip3dSig();
+	if(flip == true){
+	  TrackSip2dVal = -1.0*TrackSip2dVal;
```
```- TrackSip2dVal;``` should be good enough

emilbols replied:
Done in the new PR at https://github.com/cms-sw/cmssw/pull/23467/files#diff-898f4df86b3e380a58ff390ca3e02194R23

slava77 commented on RecoBTag/DeepFlavour/plugins/DeepFlavourTagInfoProducer.cc:
```diff
@@ -300,12 +305,13 @@ void DeepFlavourTagInfoProducer::produce(edm::Event& iEvent, const edm::EventSet
       auto entry = c_sortedindices.at(i);
       // get cached track info
       auto & trackinfo = trackinfos.at(i);
+      if(flip_ && (trackinfo.getTrackSip3dSig() > 10.0)){continue;}
```
shouldn't this be configurable like in ```pfDeepCSVNegativeTagInfos.computer.trackSelection.sip3dSigMax``` and related parameters?

emilbols replied:
Yes it should be. In fact it should always be the same as pfDeepCSVNegativeTagInfos.computer.trackSelection.sip3dSigMax since deepflavour inherits some tagging variables from deepcsv. If its not the same there will be a mismatch in cuts between some of deep flavour tagging variables.

slava77 replied:
OK.
A more strict solution is to use the provenance of the pfDeepCSVNegativeTagInfos input.

```c++
    const edm::Provenance *prov = shallow_tag_infos.provenance();
    edm::ParameterSet psetFromProvenance = edm::parameterSet(*prov);
//get the parameter values here
```

emilbols replied:
Done in the new PR at https://github.com/cms-sw/cmssw/pull/23467/files#diff-778e7825bed1fc0ad81083affff99fe3R277

slava77 commented on RecoBTag/SecondaryVertex/plugins/TemplatedSecondaryVertexProducer.cc:
```diff
@@ -860,9 +860,10 @@ void TemplatedSecondaryVertexProducer<IPTI,VTX>::produce(edm::Event &event,
 		      for(size_t iExtSv = 0; iExtSv < extSecVertex->size(); iExtSv++){
 			 const VTX & extVertex = (*extSecVertex)[iExtSv];
-			 if( Geom::deltaR2( ( position(extVertex) - pv.position() ), jetDir ) > extSVDeltaRToJet*extSVDeltaRToJet || extVertex.p4().M() < 0.3 )
+			 if( Geom::deltaR2( ( position(extVertex) - pv.position() ), (extSVDeltaRToJet > 0) ? jetDir : -jetDir ) > extSVDeltaRToJet*extSVDeltaRToJet || extVertex.p4().M() < 0.3 )
```
this looks like a bugfix that affects more than just Deep\* negative taggers.
It's worth to be noted in the PR description.

emilbols replied:
What do you mean here? That it also affects CSV?

slava77 replied:
I'm re-reading the description and now I see that it also mentions "(and CSVv2) ".
OK then.

slava77 commented on RecoVertex/AdaptiveVertexFinder/plugins/InclusiveVertexFinder.h:
```diff
@@ -285,11 +285,11 @@ void TemplatedInclusiveVertexFinder<InputContainer,VTX>::produce(edm::Event &eve
-			//                        std::cout << "Vscal: " <<  vscal << std::endl;
-			if(dlen.significance() > vertexMinDLenSig  && vscal > vertexMinAngleCosine &&  v->normalisedChiSquared() < 10 && dlen2.significance() > vertexMinDLen2DSig)
+			float vscal = dir.unit().dot((sv-ppv).unit());
+			if(dlen.significance() > vertexMinDLenSig  && ( (vertexMinAngleCosine > 0) ? (vscal > vertexMinAngleCosine) : (vscal < vertexMinAngleCosine) ) &&  v->normalisedChiSquared() < 10 && dlen2.significance() > vertexMinDLen2DSig)
```
please add some newlines to make this complex conditional more readable, esp the part related to ```(vertexMinAngleCosine > 0) ? exprA : exprB ```

emilbols replied:
Done in https://github.com/cms-sw/cmssw/pull/23206/commits/2b642284a3f3602e2e206c6684c1397ea3748e73

slava77 commented on RecoVertex/AdaptiveVertexFinder/python/inclusiveNegativeVertexing_cff.py:
```diff
@@ -0,0 +1,62 @@
+
+
+inclusiveCandidateNegativeVertexFinder = inclusiveCandidateVertexFinder.clone(
+   vertexMinAngleCosine = cms.double(-0.95) 
```
here and in other ```x = y.clone( par = ``` cases please drop the type specifications on the right hand side for changes in existing parameter values.
This is a more safe way against syntax mistakes

emilbols replied:
Done in https://github.com/cms-sw/cmssw/pull/23206/commits/4e37f178d6a3ee3b5778ba933e82064eaf694aa4

I did the same changes in inclusiveVertexing_cff.py in https://github.com/cms-sw/cmssw/pull/23206/commits/4666d808010a3b25a37ebb86fec277bb3e0e529e

And the same changes in RecoBTag/SecondaryVertex/python/pfInclusiveSecondaryVertexFinderNegativeTagInfos_cfi.py with https://github.com/cms-sw/cmssw/pull/23206/commits/575defa79289ec3f2b9bd27f3767e3defa0f3a08

btw do you know if there is a better syntax for https://github.com/cms-sw/cmssw/pull/23206/commits/4e37f178d6a3ee3b5778ba933e82064eaf694aa4#diff-8286bd0001c5e9e542e42a3b99496ad5R10 . It seems a bit clunky to call clone inside the clone, but if you do cms.PSet it resets to default values rather than cloning.

makortel replied:
@emilbols You can do
```python
inclusiveCandidateNegativeVertexFinder = inclusiveCandidateVertexFinder.clone(
   vertexMinAngleCosine = -0.95, 
   clusterizer = dict(
        clusterMinAngleCosine = -0.5
        )
)
```

and
```python
pfDeepCSVNegativeTagInfos = pfDeepCSVTagInfos.clone(
	computer = dict(
			trackSelection = dict(
				sip3dSigMax = 10.0
			)
		)
	)
```

emilbols replied:
@makortel Thanks!

slava77 commented on RecoVertex/AdaptiveVertexFinder/python/inclusiveNegativeVertexing_cff.py:
```diff
@@ -0,0 +1,62 @@
+inclusiveCandidateNegativeVertexFinder.clusterizer.clusterMinAngleCosine = cms.double(-0.5)
+
+candidateNegativeVertexMerger = candidateVertexMerger.clone()
+candidateNegativeVertexMerger.secondaryVertices = cms.InputTag("inclusiveCandidateNegativeVertexFinder")
```
keep this (and other similar cases) in the .clone method arguments

emilbols replied:
Done in https://github.com/cms-sw/cmssw/pull/23206/commits/4e37f178d6a3ee3b5778ba933e82064eaf694aa4

slava77 commented on RecoVertex/AdaptiveVertexFinder/python/inclusiveNegativeVertexing_cff.py:
```diff
@@ -0,0 +1,62 @@
+inclusiveCandidateNegativeSecondaryVertices = candidateVertexMerger.clone()
+inclusiveCandidateNegativeSecondaryVertices.secondaryVertices = cms.InputTag("candidateNegativeVertexArbitrator")
+inclusiveCandidateNegativeSecondaryVertices.maxFraction = 0.2
+inclusiveCandidateNegativeSecondaryVertices.minSignificance = 10.
```
move inside .clone( arguments

emilbols replied:
Done in https://github.com/cms-sw/cmssw/pull/23206/commits/4e37f178d6a3ee3b5778ba933e82064eaf694aa4

slava77 reviewed (commented):
IIRC, the negative taggers are not a part of the standard reco/miniAOD processing.
OTOH, the update in the DeepFlavour training is a part of the standard processing.

The overlaps in #23033 are only in the DeepFlavour negative taggers.

May I ask to decouple the bugfixes in the negative taggers from the DeepFlavour training. This will simplify the integration of the changes that will run in production from the rest.
A bit more complicated decoupling from #23033 would be to split off the negative tagger fixes in DeepFlavour package.

slava77 commented:
@cmsbuild please test

the external should be now available in the most recent IB

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 36b1cc67e21144c9139671d4155093aa403c49b7
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23206/28255/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/f519fc5608cb96c92b5768cb699fd1fcd5314cdb
https://github.com/cms-sw/cmssw/commit/7f15f5baf91fbddecf178649162f80d435067acd
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23206/28255/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23206/28255/git-merge-result

cmsbuild commented:
Comparison job queued.

emilbols committed:
Remove DeepFlavour Negative Tagger

emilbols committed:
Add newlines

emilbols committed:
move arguments inside .clone

emilbols committed:
more arguments into .clone

emilbols committed:
move inside .clone

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23206/4895

cmsbuild commented:
Pull request #23206 was updated. @perrotta, @monttj, @cmsbuild, @slava77, @gpetruc, @arizzi can you please check and sign again.

emilbols commented:
I have changed this PR so it just fixes the existing bugs in DeepCSV, so this PR no longer adds the deepflavour negative tagger and the new deepflavour training. I have also addressed the comments by @slava77 that are still relevant to this PR after the refactoring.

emilbols committed:
use dict()

emilbols committed:
use dict()

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23206/4897

cmsbuild commented:
Pull request #23206 was updated. @perrotta, @monttj, @cmsbuild, @slava77, @gpetruc, @arizzi can you please check and sign again.

ferencek commented:
Please note that having a set of commits that introduces new features followed by a set of commits that reverts them as part of a PR is generally a bad idea. It unnecessarily pollutes the official code history and on top of that makes rebasing potentially quite painful if merge conflicts appear anywhere along the commit history. With the latest changes the PR name and description should in principle also be updated in order not to be misleading.

emilbols commented:
Updated the description and title of this PR to reflect the recent changes

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 2d46d4300e5c37987cf51a4b9b9628f09cd28fb3
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23206/28325/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/f519fc5608cb96c92b5768cb699fd1fcd5314cdb
https://github.com/cms-sw/cmssw/commit/7f15f5baf91fbddecf178649162f80d435067acd
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23206/28325/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23206/28325/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23206/28325/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2902471
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2902279
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 128 log files, 14 edm output root files, 31 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/23206	2d46d43
- bugfixes in negative taggers (not including fixes in DeepFlavour, due to pending overlap ordering choice wrt #23033 ). Changes look reasonable. The updates in this PR do not affect taggers running in standard workflows.
- jenkins tests pass and comparisons with the baseline show no differences.

fabiocos commented:
@emilbols @slava77 as far as I can see from my check, this PR and #23033 are technical independent, so the order in which they are merged does not matter. Could you please confirm?

slava77 commented:
On 6/1/18 2:22 AM, Fabio Cossutti wrote:
> @emilbols <https://github.com/emilbols> @slava77
> <https://github.com/slava77> as far as I can see from my check, this PR
> and #23033 <https://github.com/cms-sw/cmssw/pull/23033> are technical
> independent, so the order in which they are merged does not matter.
> Could you please confirm?
> 
> —

Thank you for checking.
I meant to say already that this is independent in the signoff message,
but I was apparently not clear enough.

fabiocos commented:
+1

fabiocos commented:
merge
